### PR TITLE
[SYCLomatic] Refine min/max migration

### PR DESCRIPTION
**Pull request by zhiweij1**
Signed-off-by: Jiang, Zhiwei <zhiwei.jiang@intel.com>

### CODE REVIEW

File: clang/test/dpct/math-functions.cu [hunk 1]
```diff
@@ -640,3 +640,24 @@ __global__ void foo4() {
...
 }
+
+struct A {};
+
+__device__ A min(A a, A b) { return a; }
+__device__ A max(A a, A b) { return a; }
+
+template <class T> __device__ T clamp(T x, T a, T b) {
+  // CHECK: /*
+  // CHECK-NEXT: DPCT1064:{{[0-9]+}}: Migrated min call is used in a macro/template definition and is not valid for all macro/template uses. Adjust the code.
```
Comment (zhiweij1) [inline clang/test/dpct/math-functions.cu]:
```suggestion
  // CHECK-NEXT: DPCT1064:{{[0-9]+}}: Migrated min call is used in a macro/template definition and may not be valid for all macro/template uses. Adjust the code.
```

File: clang/test/dpct/macro_test.cu [hunk 2]
```diff
@@ -520,7 +520,7 @@ __global__ void foo8(){
 
 
 //CHECK: #define DFABS(x) (double)sycl::fabs((x))
-//CHECK-NEXT: #define MAX(x, y) sycl::max(x, y)
+//CHECK-NEXT: #define MAX(x, y) dpct::max(a, sycl::sqrt(DFABS(b)))
```
Comment (zhiweij1) [inline clang/test/dpct/macro_test.cu]: Mark this checker, it is a regression

Comment (zhiweij1) [reply clang/test/dpct/macro_test.cu]: And check why using dpct here. Should be sycl?

File: clang/test/dpct/cuda_kernel_include.cu [hunk 3]
```diff
@@ -10,8 +10,8 @@
 
 // CHECK: void hello(const sycl::nd_item<3> &item_ct1) {
 // CHECK-NEXT:  int index = item_ct1.get_group(2) * item_ct1.get_local_range(2) + item_ct1.get_local_id(2);
-// CHECK-NEXT:  int tmp = sycl::min((unsigned int)((item_ct1.get_group(2)+1)*item_ct1.get_local_range(2)+item_ct1.get_local_id(2)), (unsigned int)(sycl::max(index, 45)));
-// CHECK-NEXT:  int num = sycl::max((unsigned int)((item_ct1.get_group(2)+1)*item_ct1.get_local_range(2)+item_ct1.get_local_id(2)), (unsigned int)(sycl::min(tmp, 45)));
+// CHECK-NEXT:  int tmp = dpct::min((item_ct1.get_group(2)+1)*item_ct1.get_local_range(2)+item_ct1.get_local_id(2), sycl::max(index, 45));
```
Comment (zhiweij1) [inline clang/test/dpct/cuda_kernel_include.cu]: Should we keep the cast to unsigned int ?

File: clang/test/dpct/builtin_warpSize.cu [hunk 4]
```diff
@@ -18,7 +18,7 @@ __global__ void foo(){
 }
 
 // CHECK: void bar(const sycl::nd_item<3> &item_ct1){
-// CHECK-NEXT:   int a = sycl::max((int)item_ct1.get_sub_group().get_local_range().get(0), 0);
+// CHECK-NEXT:   int a = dpct::max((int)item_ct1.get_sub_group().get_local_range().get(0), 0);
```
Comment (zhiweij1) [inline clang/test/dpct/builtin_warpSize.cu]: should here use sycl version?

File: clang/test/dpct/macro_test.cu (hunk 2 above)
Comment (zhiweij1) [reply clang/test/dpct/macro_test.cu]: > And check why using dpct here. Should be sycl?

In `math::UsingDpctMinMax`, we check if this API is macro. If it is macro, we use `dpct::max/dpct::min`.

File: clang/test/dpct/builtin_warpSize.cu (hunk 4 above)
Comment (zhiweij1) [reply clang/test/dpct/builtin_warpSize.cu]: fixed

File: clang/test/dpct/math-functions.cu (hunk 1 above)
Comment (zhiweij1) [reply clang/test/dpct/math-functions.cu]: fixed

File: clang/test/dpct/cuda_kernel_include.cu (hunk 3 above)
Comment (zhiweij1) [reply clang/test/dpct/cuda_kernel_include.cu]: fixed

File: clang/test/dpct/macro_test.cu (hunk 2 above)
Comment (zhiweij1) [reply clang/test/dpct/macro_test.cu]: > Mark this checker, it is a regression

Done

**Review by zhimingwang36 — APPROVED** [review]
LGTM

**Pull request merged**